lib/datadog/appsec/contrib/rack/reactive/request.rb

```diff
@@ -38,10 +38,15 @@ def self.subscribe(op, waf_context)
                 cookies = values[3]
                 client_ip = values[4]
 
+                # rules such as coming from the passlist feature assume a relative URI
+                if uri_raw && (m = %r{^(?<scheme>[a-z]+)://(?<authority>[^/]+)?(?<relative_uri>/.*)}.match(uri_raw))
+                  relative_uri = m['relative_uri']
+                end
+
                 waf_args = {
                   'server.request.cookies' => cookies,
                   'server.request.query' => query,
-                  'server.request.uri.raw' => uri_raw,
+                  'server.request.uri.raw' => relative_uri,
```

Comment:
I suggest moving this logic to `Gateway::Request` object

Rack::Request have a method call `:fullpath` that actually returns what we want:
https://github.com/rack/rack/blob/main/lib/rack/request.rb#L592

So after adding the method `fullpath` to `Gateway::Request` object, we can use it in:

```ruby
def self.publish(op, gateway_request)
  catch(:block) do
    op.publish('request.query', gateway_request.query)
    op.publish('request.headers', gateway_request.headers)
    op.publish('request.uri.raw', gateway_request.fullpath)
    op.publish('request.cookies', gateway_request.cookies)
    op.publish('request.client_ip', gateway_request.client_ip)
    nil
  end
end
```

This has the added benefit that we do not use a custom regex in the critical path, and that we use the already provided `Rack::Request` methods

Reply:
> I suggest moving this logic to Gateway::Request object

Now that specs are passing, I'm attempting to do just that.

Sadly `fullpath` relies on parsing the URL and even _recomposes_ it from `path` + `query_string` parsed elements, and our internal address RFC mandates that the `server.request.uri` address MUST NOT be parsed.

```
      def fullpath
        query_string.empty? ? path : "#{path}?#{query_string}"
      end
```

Reply:
Previously we were using `url` methods, which also recomposes from `base_url` and `fullpath`

```ruby
def url
  base_url + fullpath
end
```

So I don't see the difference or issue to just using `fullpath`

Reply:
> We were using already parsed values before

Ah, correct, we were using `request.url`, which is actually made from `fullpath`:

```
      def url
        base_url + fullpath
      end
```

Reply:
> We were using already parsed values before

Ah, correct, we were using [`request.url`](https://github.com/rack/rack/blob/54990d39446caf046b5cbf19fd030204f850b480/lib/rack/request.rb#L584-L586), which is actually made from `fullpath`:

```
      def url
        base_url + fullpath
      end
```

Reply:
A note that we have another parser here which attempts to use `Request-URI` instead: https://github.com/DataDog/dd-trace-rb/blob/9bebd8c84f4743eab535c2ce4f06289e8899812c/lib/datadog/tracing/contrib/rack/middlewares.rb#L271-L311